File: src/app/conjugation/forms/past-short-affirmative.ts
import {Conjugation, Word, WordType} from "../conjugation";
import {TeFormAffirmative} from "./te-form-affirmative";

export class PastShortAffirmative implements Conjugation {
    getConjugation(word: Word): (Word | undefined) {
        switch (word.wordType) {

            // Verbs
            case WordType.GodanVerb:
            case WordType.IchidanVerb:
            case WordType.SuruVerb:
            case WordType.KuruVerb:
                const teForm = new TeFormAffirmative().getConjugation(word)
                if (teForm === undefined) {
                    return undefined
                }

                if (teForm.getLastKana() === 'で') {
                    return teForm.replaceLastKana('だ')
                } else {
                    return teForm.replaceLastKana('た')
                }


            // Adjectives
            case WordType.IAdjective:
                return word.replaceLastKana('かった')

            case WordType.NaAdjective:
                return word.addSuffix('だった')


            default:
                return undefined;
        }
    }

    getTitle(): string {
        return "Past, short, affirmative"
    }

    getSettingsTitle(): string {
        return "Short, affirmative"
    }

}

File: backend/app/conjugation/forms/past_short_affirmative.py
from ..core import Conjugation, Word, WordType
from .te_form_affirmative import TeFormAffirmative


class PastShortAffirmative(Conjugation):

    title = 'Past, short, affirmative'
    settings_title = 'Short, affirmative'

    def conjugate(self, word: Word) -> Word | None:
        match word.word_type:

            # Verbs
            case (
                WordType.GODAN_VERB
                | WordType.ICHIDAN_VERB
                | WordType.SURU_VERB
                | WordType.KURU_VERB
            ):
                te_form = TeFormAffirmative().conjugate(word)
                if te_form is None:
                    return None

                if te_form.last_kana == 'で':
                    return te_form.replace_last_kana('だ')
                else:
                    return te_form.replace_last_kana('た')

            # Adjectives
            case WordType.I_ADJECTIVE:
                return word.replace_last_kana('かった')

            case WordType.NA_ADJECTIVE:
                return word.add_suffix('だった')

            case _:
                return None
